### PR TITLE
Replace index.d.ts with generated type definitions

### PR DESCRIPTION
**Pull request by colebemis**
## Problem

Primer React was originally released without TypeScript type definitions, making it difficult for engineers to consume the library in TypeScript applications. In [July 2019](https://github.com/primer/components/commit/2983c935ea9ad600c04078adb25e40c3624c11fa#diff-7aa4473ede4abd9ec099e87fec67fd57afafaf39e05d493ab4533acc38547eb8), we created an [ambient declaration](https://www.geeksforgeeks.org/typescript-ambients-declaration/) file (`index.d.ts`) file to provide type definitions for TypeScript applications without having to rewrite Primer React components in TypeScript. 

`index.d.ts` has been an effective stopgap, enabling teams to build complex applications with Primer React and TypeScript. However, because `index.d.ts` is disconnected from the implementation code, we've struggled to keep the type definitions up-to-date and accurate, as evidenced by [many](https://github.com/primer/components/issues/906) [TypeScript](https://github.com/primer/components/issues/540) [bug](https://github.com/primer/components/issues/520) [reports](https://github.com/primer/components/issues/534). As the library continues to grow in size and complexity, manually maintaining type definitions will become unsustainable.

## Solution

Now that all our components have been rewritten in TypeScript, we can use the TypeScript compiler to generate type definitions instead of manually maintaining an ambient declaration file. This PR removes our manually maintained `index.d.ts` file and updates the `package.json` to reference the generated type definitions.

## Checklist

- [x] Export prop types from `index.ts` https://github.com/primer/components/pull/1147/commits/21c806cd66d1b4609600e043c8300ed8e0458b52
- [x] Update `typings` field in `package.json` to reference generated type definitions https://github.com/primer/components/pull/1147/commits/bab3e96ff16359e42feba8b013f27a892f3157f1
- [x] Test generated type definitions in an application
- [x] Fix type errors
- [x] Delete `index.d.ts`
- [x] Write changeset

### CODE REVIEW

File: tsconfig.json [hunk 1]
```diff
@@ -6,7 +6,7 @@
     "checkJs": false,
     "jsx": "preserve",
     "declaration": true,
-    "outDir": "./types",
+    "outDir": "./lib",
```
Comment (dgreif) [inline tsconfig.json:9]: Looks like we can remove the `gitignore` for `types` based on this change.

**Review by dgreif — COMMENTED** [review]
Sorry, didn't realize this was a Draft PR before I started reviewing 😄 .  Ignore anything you already have fixed up

File: tsconfig.json (hunk 1 above)
Comment (colebemis) [reply tsconfig.json:9]: Yes. Good catch!

File: src/index.ts [hunk 2]
```diff
@@ -1,14 +1,21 @@
 export {default as theme} from './theme'
 export {get as themeGet} from './constants'
 export {default as BaseStyles} from './BaseStyles'
+export type {BaseStylesProps} from './BaseStyles'
```
[the remark this answers is not in the document]
Comment (colebemis) [reply src/index.ts:4]: If they're on the same line, Babel won't remove the type import during compilation which causes errors like "BaseStylesProps is not defined".

File: src/index.ts [hunk 3]
```diff
@@ -35,38 +47,108 @@ export {
...
+  TimelineBodyProps,
+  TimelineBreakProps,
+  TimelineItemsProps
+} from './Timeline'
 export {default as Tooltip} from './Tooltip'
+export type {TooltipProps} from './Tooltip'
 export {default as Truncate} from './Truncate'
+export type {TruncateProps} from './Truncate'
 export {default as UnderlineNav} from './UnderlineNav'
+export type {UnderlineNavProps, UnderlineNavLinkProps} from './UnderlineNav'
```
[the remark this answers is not in the document]
Comment (colebemis) [reply src/index.ts:154]: Thanks for checking this! These are intentionally excluded but I'll add a note in the release notes.

**Review by dgreif — APPROVED** [review]
Awesome work, this is a major accomplishment! 🏆 🚀

**Pull request merged**